core/src/main/java/com/linecorp/armeria/internal/common/util/KeyStoreUtil.java

```diff
@@ -0,0 +1,186 @@
+                        throw newException("the key pair contains no certificate chain", keyStoreFile,
+                                           "(Specify the alias to choose the right key pair or" +
+                                           " ensure the key pair has a certificate chain.");
+                    }
+
+                    if (privateKey != null) {
+                        throw newException("found more than one key pair from key store", keyStoreFile,
+                                           "(Specify the alias to choose the right key pair or" +
+                                           " use the key store that has only one key pair.)");
+                    }
+
+                    privateKey = candidateKey;
+                    certificateChain = Arrays.stream(candidateCertificateChain)
+                                             .map(X509Certificate.class::cast)
+                                             .collect(ImmutableList.toImmutableList());
+                }
+            }
+
+            if (privateKey == null) {
+                throw newException("no key pair found from key store", keyStoreFile,
+                                   "(Use the key store that has at least one key pair.)");
+            }
+
+            assert certificateChain != null;
+
+            return new KeyPair(privateKey, certificateChain);
+        }
+    }
+
+    @Nullable
+    private static String detectFormat(InputStream keyStoreStream) throws IOException {
+        final byte[] magic = new byte[4];
+        try {
+            ByteStreams.readFully(keyStoreStream, magic);
+        } catch (EOFException unused) {
+            // The file is shorter than 4 bytes.
+            return null;
+        }
+
+        if (magic[0] == 0x30 && magic[1] == (byte) 0x82) {
+            return "PKCS12";
+        }
+
+        if (magic[0] == (byte) 0xFE && magic[1] == (byte) 0xED &&
+            magic[2] == (byte) 0xFE && magic[3] == (byte) 0xED) {
+            return "JKS";
+        }
+
+        return null;
+    }
+
+    @Nullable
+    private static char[] keyPassword(@Nullable String keyStorePassword, @Nullable String keyPassword) {
+        if (keyPassword != null) {
+            return keyPassword.toCharArray();
+        }
+        if (keyStorePassword != null) {
+            return keyStorePassword.toCharArray();
+        }
+        return null;
+    }
+
+    private static IllegalArgumentException newException(String message, @Nullable File keyStoreFile,
+                                                         String additionalInfo) {
+        if (keyStoreFile != null) {
+            return new IllegalArgumentException(
+                    message + ": " + keyStoreFile + ' ' + additionalInfo);
+        } else {
+            return new IllegalArgumentException(message + ' ' + additionalInfo);
+        }
+    }
+
+    private KeyStoreUtil() {}
+
+    public static final class KeyPair {
```

Comment:
`java.security.KeyPair`, which has just a public key, is widely used so I'm worried that this might confuse users because it's the same name. But couldn't think of a better name so fine as is.

Reply:
`X509KeyPair` 🤔 ?

Reply:
This is an internal utility class, so it will never be exposed in our public API. Let me rename if necessary when I actually incorporate into our public API.